lib/src/message.dart

```diff
@@ -59,16 +58,30 @@ abstract class Message {
   /// Content-Type header, it will be set to "application/octet-stream".
   Message(body,
       {Encoding encoding,
-      Map<String, String> headers,
-      Map<String, Object> context})
-      : this._(new Body(body, encoding), headers, context);
+        Map<String, String> headers,
+        Map<String, Object> context})
+      : this.__(body, _determineMediaType(body, encoding, headers), headers,
+      context);
 
-  Message._(Body body, Map<String, String> headers, Map<String, Object> context)
+  Message.__(body, MediaType contentType, Map<String, String> headers,
+      Map<String, Object> context)
+      : this._(new Body(body, encodingForMediaType(contentType, null)), contentType,
+      headers, context);
+
+  Message._(Body body, MediaType contentType, Map<String, String> headers,
+      Map<String, Object> context)
+      : this.fromValues(body, _adjustHeaders(headers, body, contentType), context);
+
+  /// Creates a new [Message].
+  ///
+  /// This constructor should be used when no computation is required for the
+  /// [body], [headers] or [context].
+  Message.fromValues(
```

Comment:
I understand the performance motivation for having this constructor, but I'd like to focus on API shape and correctness for now and leave optimizations like this for later (ideally driven by real data).

Reply:
Ok I'll rework this accordingly.